src/python/pants/engine/internals/rule_visitor.py

```diff
@@ -28,6 +36,102 @@ def _get_starting_indent(source: str) -> int:
     return 0
 
 
+def _node_str(node: Any) -> str:
+    if isinstance(node, ast.Name):
+        return node.id
+    if isinstance(node, ast.Attribute):
+        return ".".join([_node_str(node.value), node.attr])
+    if isinstance(node, ast.Call):
+        return _node_str(node.func)
+    if sys.version_info[0:2] < (3, 8):
+        if isinstance(node, ast.Str):
+            return node.s
+    else:
+        if isinstance(node, ast.Constant):
+            return str(node.value)
+    return str(node)
+
+
+class _TypeStack:
+    def __init__(self, func: Callable) -> None:
+        self._stack: list[dict[str, Any]] = []
+        self.root = sys.modules[func.__module__]
+        self.push(self.root)
+        self.__push_function_closures(func)
+
+    def __repr__(self) -> str:
+        from pprint import pformat
+
+        return f"TypeStack<\n{pformat(self._stack)}\n>"
+
+    def __getitem__(self, name: str) -> Any:
+        for ns in reversed(self._stack):
+            if name in ns:
+                return ns[name]
+        return self.root.__builtins__.get(name, None)
+
+    def __setitem__(self, name: str, value: Any) -> None:
+        self._stack[-1][name] = value
+
+    def __push_function_closures(self, func: Callable) -> None:
```

Comment:
Avoid mangled method names if you can :smile:

Reply:
do you mean avoid the double underscore?
I don't mind using a single one to avoid the "private" Python thing that tacks the class name onto it.

Reply:
Yes please. usually when I see dunder in front I get suspicious and wonder if/why we need mangling.

Reply:
fair enough. I just read "private".. but there's really no such thing in Python anyway, only "slightly harder to mess with".